clippy_lints/src/legacy_numeric_constants.rs

```diff
@@ -0,0 +1,227 @@
+use clippy_utils::{
+    diagnostics::span_lint_and_then,
+    get_parent_expr, is_from_proc_macro, last_path_segment,
+    msrvs::{self, Msrv},
+};
+use itertools::Itertools;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_errors::Applicability;
+use rustc_hir::{
+    def::Res,
+    def_id::DefId,
+    intravisit::{walk_expr, Visitor},
+    Item, UseKind,
+};
+use rustc_hir::{Expr, ExprKind, ItemKind, PrimTy, QPath, TyKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::{hir::nested_filter::OnlyBodies, lint::in_external_macro};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{sym, Span, Symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `<integer>::max_value()`, `std::<integer>::MAX`,
+    /// `std::<float>::EPSILON`, etc.
+    ///
+    /// ### Why is this bad?
+    /// All of these have been superceded by the associated constants on their respective types,
+    /// such as `i128::MAX`. These legacy constants may be deprecated in a future version of rust.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let eps = std::f32::EPSILON;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let eps = f32::EPSILON;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub LEGACY_NUMERIC_CONSTANTS,
+    style,
+    "checks for usage of legacy std numeric constants and methods"
+}
+pub struct LegacyNumericConstants {
+    msrv: Msrv,
+    use_stmts: FxHashMap<Symbol, Vec<Span>>,
+    glob_use_stmts: Vec<Span>,
+}
+
+impl LegacyNumericConstants {
+    #[must_use]
+    pub fn new(msrv: Msrv) -> Self {
+        Self {
+            msrv,
+            use_stmts: FxHashMap::default(),
+            glob_use_stmts: vec![],
+        }
+    }
+}
+
+impl_lint_pass!(LegacyNumericConstants => [LEGACY_NUMERIC_CONSTANTS]);
+
+impl<'tcx> LateLintPass<'tcx> for LegacyNumericConstants {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        let Self {
+            msrv: _,
+            use_stmts,
+            glob_use_stmts,
+        } = self;
+
+        if !item.span.is_dummy() && let ItemKind::Use(path, kind) = item.kind {
+            match kind {
+                UseKind::Single => {
+                    for res in &path.res {
+                        if let Some(def_id) = res.opt_def_id()
+                            && let Some(module_name) = is_path_in_integral_module(cx, def_id)
+                            && let _ = use_stmts.insert(module_name, vec![])
+                            && let Some(use_stmts) = use_stmts.get_mut(&module_name)
+                        {
+                            use_stmts.push(item.span);
+                        }
+                    }
+                },
+                UseKind::Glob => glob_use_stmts.push(item.span),
```

Comment:
Maybe we can use `names_imported_by_glob_use` to filter some of these out? Might not be worth the effort, though